### PR TITLE
Fix import hints that are followed by dot.number

### LINKED ISSUE #78881: Import Hints are no Longer Processed if the Object Name ends with a Number
### Godot version

4.1 rc2

### System information

Manjaro GL-Compatibility

### Issue description

When using import hints, such as -col or -colonly, on object names Godot used to accept the hint even if it was followed by Blender's auto duplicate naming scheme such as .001 or .002. This is no longer the case, items name to the likes of "object-col.001" will not import with collision.

I noticed that, prior to 4.1, Godot would remove periods from imported object names. Object.001 would import as Object001, but now it would import as Object_001. Perhaps this change broke the import hint behavior.

### Steps to reproduce

Export a scene with object names using import hints, such as -col or -convcol, where some of the objects have '.001' or similar appended to the end of their names. Import the scene into Godot and create an inherited scene.

### Minimal reproduction project

[ImportHints.zip](https://github.com/godotengine/godot/files/11919131/ImportHints.zip)

### PR DESCRIPTION
**Pull request by capnm**
Fixes #78881.

### CODE REVIEW

**Comment by dalexeev** [comment]
Could you please also update the docs: `String.validate_node_name()`, `StringName.validate_node_name()`?

> Returns a copy of the string with all characters that are not allowed in Node.name removed (. : @ / " %).

```gdscript
# 4.0.3: 123
# master: 1_2_3
print("1.2@3".validate_node_name())
```

This is not directly related to the PR title, but it is also a consequence of #75760.

File: editor/import/resource_importer_scene.cpp [hunk 1]
```diff
@@ -313,7 +313,7 @@ static bool _teststr(const String &p_what, const String &p_str) {
 	String what = p_what;
 
 	//remove trailing spaces and numbers, some apps like blender add ".number" to duplicates so also compensate for this
-	while (what.length() && (is_digit(what[what.length() - 1]) || what[what.length() - 1] <= 32 || what[what.length() - 1] == '.')) {
+	while (what.length() && (is_digit(what[what.length() - 1]) || what[what.length() - 1] <= 32 || what[what.length() - 1] == '_')) {
```
Comment (akien-mga) [inline editor/import/resource_importer_scene.cpp:316]: GH won't let you comment on the right line but: The comment should be amended to clarify, it's talking about `.number` and we're handling `_number`.

I'm also a bit surprised that the fix is this, reduz's PR changed it to replace `.` by `_` instead of an empty string. But here it sounds like the previous logic still handled `.` so it happened _before_ the replacement, and now it seems to happen after?

Comment (capnm) [reply editor/import/resource_importer_scene.cpp:316]: I think he did it right :) I can just replace the old `hint.num3er` with `hint_num3er`.

Comment (capnm) [reply editor/import/resource_importer_scene.cpp:316]: ?
// Remove trailing spaces and numbers, some apps like blender add ".number" 
// (dot is replaced by _ as invalid character) to duplicates so also compensate for this.

Comment (akien-mga) [reply editor/import/resource_importer_scene.cpp:316]: Sounds good.

**Comment by akien-mga** [comment]
So I did some testing, it seems like it's a decent workaround for the current regression, though this warrants further work to make things more robust.

The key change here is that this logic used to run _before_ validating the node names, so we could properly handle the `.001` from Blender. Now that it runs after the rename, we can't differentiate between `.001` and `_001` coming from Blender. So e.g. in 4.0.3, a node named `Cube-convcolonly_001` would not be handled as having the `-convcolonly` hint, while with this change it will.

To test this I made this scene derived from the MRP of #78881:

image

### 4.0.3

Here's how it imports in 4.0.3 using different formats (.blend file directly, .glb exported from Blender 3.6, .dae exported from Blender 3.6):

#### blend

image

Behaves as expected for the 4.0.3 logic, the `.001` and `.002` hints are properly handled. The one I added with `_002` which is invalid syntax for an import hint is left untouched, so it's imported as a normal cube and keeps its name.

The `Cube-convcolonly` which conflicts with the existing `Cube` gets auto-renamed to solve the naming conflict.

#### glb

image

Similar behavior as blend, minus the light and camera but that's properly from not setting the right export settings, and unrelated to this issue.

#### dae

image

Also similar result. The scene tree order seems backwards, but that might be a bug in the Collada pipeline which we don't actively maintain.

### 4.1 RC 3

Reproducing the bug first to be able to validate the change.

#### blend

image

As expected, hints aren't properly handled, aside from the one on `Cube-convcolonly`.

Worth noting that the node name disambiguation logic changed too, so it's no longer named `@Cube@...` but it's actually using the node type instead with `@StaticBody3D@...`. This may or may not be another intentional change from #75760 (judging by the code I would say it's intentional and part of the optimization).

#### glb

image

Consistent with blend.

#### dae

image

Consistent with above results too. Note how the name conflict handling differs between the glTF and collada pipelines (duplicate `Cube-convcolonly_002` becomes respectively `Cube-convcolonly_0022` for glTF and `Cube-convcolonly_003` for collada).

### This PR

Confirming that it fixes the issue on the "expected" case of having `.001` suffix, and creates a minor bug with now supporting `_001` as a valid suffix to trim too. This is a slight breakage, but I think acceptable to fix the regression in the short term.

#### blend

image

As expected, fixes the bug, creates a minor behavior change for `Cube-convcolonly_002` which is now handled as a `-convcolonly` hint. Technically a bug, could be considered a feature ;)

#### glb

image

Same result.

#### dae

image

Also working as expected.

---

**TL;DR:** It's not perfect, but I think it's good enough to YOLO merge now for 4.1 and fix the regression. I'd appreciate that the @godotengine/import teams looks further into it afterwards anyway, and makes sure that this system is more robust.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
@capnm Could you rebase your branch? The current version is 800+ commits behind `master`, would be good to reduce the diff.

**Comment by capnm** [comment]
> @capnm Could you rebase your branch? The current version is 800+ commits behind master, would be good to reduce the diff.

I forgot to rebase the github repo.

> **TL;DR:** It's not perfect, but I think it's good enough to YOLO merge now for 4.1 and fix the regression. I'd appreciate that the @godotengine/import teams looks further into it afterwards anyway, and makes sure that this system is more robust.

It is a good idea from old times, but changing names is very limited and as you can see very bad to handle. Now the Blender glTF exporter is able to export custom object properties, it is IMO much better to use that.

image

```
"extras":{
	"prop_capnm":24.329999923706055,
	"capnm_str":"Das ist mein string mit \u00df\u00df\u00df!"
},
```



For example, this doesn't look fast...
```
[L] GD (./editor/import/resource_importer_scene.cpp:333) _teststr:
return FALSE: p_what '' p_str 'colonly' what ''
[L] GD (./editor/import/resource_importer_scene.cpp:333) _teststr:
return FALSE: p_what '' p_str 'convcolonly' what ''
[L] GD (./editor/import/resource_importer_scene.cpp:333) _teststr:
return FALSE: p_what '' p_str 'rigid' what ''
[L] GD (./editor/import/resource_importer_scene.cpp:333) _teststr:
return FALSE: p_what '' p_str 'col' what ''
[L] GD (./editor/import/resource_importer_scene.cpp:333) _teststr:
return FALSE: p_what '' p_str 'convcol' what ''
[L] GD (./editor/import/resource_importer_scene.cpp:333) _teststr:
return FALSE: p_what '' p_str 'navmesh' what ''
[L] GD (./editor/import/resource_importer_scene.cpp:333) _teststr:
return FALSE: p_what '' p_str 'occ' what ''
[L] GD (./editor/import/resource_importer_scene.cpp:333) _teststr:
return FALSE: p_what '' p_str 'occonly' what ''
[L] GD (./editor/import/resource_importer_scene.cpp:333) _teststr:
return FALSE: p_what '' p_str 'vehicle' what ''
[L] GD (./editor/import/resource_importer_scene.cpp:333) _teststr:
return FALSE: p_what '' p_str 'wheel' what ''
```

**Review by Calinou — APPROVED** [review]
Tested locally on https://github.com/godotengine/tps-demo (rebased on top of `master` cdd2313ba27d0a2600a18e849b4c5d1fd6a6e351), it works.

**Comment by capnm** [comment]
Several seconds too late :)
